src/mds/Mutation.cc

```diff
@@ -190,24 +164,34 @@ void MutationImpl::add_cow_inode(CInode *in)
 void MutationImpl::add_cow_dentry(CDentry *dn)
 {
   pin(dn);
-  dirty_cow_dentries.push_back(pair<CDentry*,version_t>(dn, dn->get_projected_version()));
+  dirty_cow_dentries.emplace_back(dn, dn->get_projected_version());
 }
 
 void MutationImpl::apply()
 {
-  pop_and_dirty_projected_inodes();
-  pop_and_dirty_projected_fnodes();
-  
+  for (auto obj : projected_nodes) {
```

Comment:
I realize these are pointers but leaving this as a semantic copy makes a future heavier type more difficult to introduce. It's always better to see if you can do:

`for (const auto& foo : bar)`

Reply:
it's for CDir::_get_fnode() and CDir::_get_projected_fnode(). _get_fnode() is for modifying fnode of newly allocate CDir.  _get_projected_fnode() is for functions that know fnode is already projected and wants to modify it